### PR TITLE
Rewrite tokenizer to buffer input

### PR DESCRIPTION
**Pull request by ajtribick**
Results in galaxies.dsc parsing time dropping from 0.34 to 0.24 seconds on my machine

### CODE REVIEW

File: src/celutil/tokenizer.cpp [hunk 1]
```diff
@@ -8,622 +8,800 @@
...
+    double operator()(std::int32_t i) const { return static_cast<double>(i); }
+    double operator()(double d) const { return d; }
+    double operator()(std::string_view) const { return std::nan(""); }
+    double operator()(const std::string&) const { return std::nan(""); }
+};
+
+
+struct VisitorInteger
+{
+    std::optional<std::int32_t> operator()(std::monostate) const { return std::nullopt; }
```
Comment (375gnu) [inline src/celutil/tokenizer.cpp:144]: is it a good idea to provide inconsistent api for the similar purpose? option for integers and actual types for double/string_view?

File: src/celmodel/modelfile.cpp [hunk 2]
```diff
@@ -613,43 +613,61 @@ AsciiModelLoader::loadMesh(Mesh& mesh)
...
-        if (tok.nextToken() != Tokenizer::TokenNumber || !tok.isInteger())
+        if (tok.nextToken() != Tokenizer::TokenNumber)
         {
             reportError("Material index expected in primitive group");
             return false;
         }
 
         unsigned int materialIndex;
-        if (tok.getIntegerValue() == -1)
+        if (std::int32_t tokenValue = tok.getIntegerValue().value_or(-2); tokenValue == -1)
```
Comment (375gnu) [inline src/celmodel/modelfile.cpp:623]: how tokenValue can be -1 if you require to return -2?

File: src/celutil/tokenizer.cpp (hunk 1 above)
Comment (ajtribick) [reply src/celutil/tokenizer.cpp:144]: Probably not, I've updated this, plus separated out `getNameValue()` and `getStringValue()` to combine the token type tests and data retrieval

File: src/celmodel/modelfile.cpp (hunk 2 above)
Comment (ajtribick) [reply src/celmodel/modelfile.cpp]: Rewritten the condition, hopefully clearer now.

**Review by 375gnu — APPROVED** [review]

**Pull request merged**